Panic while downloading siva files to HDFS

Testing with a build done using a binary from #32 (I could not connect to HDFS otherwise), pga tool is failing to `get` files to HDFS

```
# cat /siva.txt | ./pga-rafa get --verbose -i -o hdfs://hdfs-namenode-0.hdfs-namenode.default.svc.cluster.local:8020
downloading siva files by name from stdin
filter flags will be ignored
DEBU[0000] syncing http://pga.sourced.tech//siva/latest/4a/4a14cc02da0a9280538cd3f3242365601d72f241.siva to hdfs://hdfs-namenode-0.hdfs-namenode.default.svc.cluster.local:8020/siva/latest/4a/4a14cc02da0a9280538cd3f3242365601d72f241.siva
panic: runtime error: slice bounds out of range

goroutine 1 [running]:
github.com/src-d/datasets/PublicGitArchive/pga/cmd.downloadFilenames(0x86d4e0, 0xc4201ca080, 0x86d560, 0xc4201b8030, 0xc4201dc000, 0x1f, 0x1f, 0xa, 0x8000105, 0x0)
	/root/go/src/github.com/src-d/datasets/PublicGitArchive/pga/cmd/get.go:91 +0x263
github.com/src-d/datasets/PublicGitArchive/pga/cmd.glob..func1(0xa67920, 0xc4200b6340, 0x0, 0x4, 0x0, 0x0)
	/root/go/src/github.com/src-d/datasets/PublicGitArchive/pga/cmd/get.go:79 +0x3a8
github.com/src-d/datasets/PublicGitArchive/pga/vendor/github.com/spf13/cobra.(*Command).execute(0xa67920, 0xc4200b6300, 0x4, 0x4, 0xa67920, 0xc4200b6300)
	/root/go/src/github.com/src-d/datasets/PublicGitArchive/pga/vendor/github.com/spf13/cobra/command.go:698 +0x46d
github.com/src-d/datasets/PublicGitArchive/pga/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xa67d60, 0xc4200abf58, 0x73fc95, 0xc4201763c0)
	/root/go/src/github.com/src-d/datasets/PublicGitArchive/pga/vendor/github.com/spf13/cobra/command.go:783 +0x2e4
github.com/src-d/datasets/PublicGitArchive/pga/vendor/github.com/spf13/cobra.(*Command).Execute(0xa67d60, 0xc42002a0b8, 0x0)
	/root/go/src/github.com/src-d/datasets/PublicGitArchive/pga/vendor/github.com/spf13/cobra/command.go:736 +0x2b
github.com/src-d/datasets/PublicGitArchive/pga/cmd.Execute()
	/root/go/src/github.com/src-d/datasets/PublicGitArchive/pga/cmd/root.go:34 +0x2d
main.main()
	/root/go/src/github.com/src-d/datasets/PublicGitArchive/pga/main.go:8 +0x20
```

Find attached the contents of [siva.txt](https://github.com/src-d/datasets/files/1789579/siva.txt)

For the moment I'm using multitool to download to HDFS as it is not giving me issues

cc @vmarkovtsev

--- a/PublicGitArchive/pga/cmd/filesystem.go
+++ b/PublicGitArchive/pga/cmd/filesystem.go
@@ -4,10 +4,10 @@ import (
 	"fmt"
 	"io"
 	"net/http"
+	"net/url"
 	"os"
 	"path/filepath"
 	"strconv"
-	"strings"
 	"time"
 
 	"github.com/colinmarc/hdfs"
@@ -30,18 +30,25 @@ func FileSystemFromFlags(flags *pflag.FlagSet) (FileSystem, error) {
 		return nil, err
 	}
 
-	if strings.HasPrefix(path, "http://") || strings.HasPrefix(path, "https://") {
-		return urlFS(path), nil
-	}
-	if !strings.HasPrefix(path, "hdfs://") {
-		return localFS(path), nil
+	u, err := url.Parse(path)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse output location: %v", err)
 	}
 
-	client, err := hdfs.New(strings.TrimPrefix(path, "hdfs://"))
-	if err != nil {
-		return nil, fmt.Errorf("could not create HDFS client: %v", err)
+	switch u.Scheme {
+	case "http", "https":
+		return urlFS(path), nil
+	case "hdfs":
+		client, err := hdfs.New(u.Host)
+		if err != nil {
+			return nil, fmt.Errorf("could not create HDFS client: %v", err)
+		}
+		return hdfsFS{u.Path, client}, nil
+	case "":
+		return localFS(path), nil
+	default:
+		return nil, fmt.Errorf("scheme not supported in output location %s", path)
 	}
-	return hdfsFS{path, client}, nil
 }
 
 type localFS string
@@ -112,11 +119,20 @@ type hdfsFS struct {
 	c    *hdfs.Client
 }
 
-func (fs hdfsFS) Abs(path string) string                     { return fs.path + "/" + path }
-func (fs hdfsFS) Create(path string) (io.WriteCloser, error) { return fs.c.Create(fs.Abs(path)) }
-func (fs hdfsFS) Open(path string) (io.ReadCloser, error)    { return fs.c.Open(fs.Abs(path)) }
-func (fs hdfsFS) ModTime(path string) (time.Time, error)     { return modtime(fs.c.Stat(fs.Abs(path))) }
-func (fs hdfsFS) Size(path string) (int64, error)            { return size(fs.c.Stat(fs.Abs(path))) }
+func (fs hdfsFS) Abs(path string) string { return fs.path + "/" + path }
+
+func (fs hdfsFS) Create(path string) (io.WriteCloser, error) {
+	path = fs.Abs(path)
+	dir := filepath.Dir(path)
+	if err := fs.c.MkdirAll(dir, os.ModePerm); err != nil && !os.IsExist(err) {
+		return nil, fmt.Errorf("could not create %s: %v", dir, err)
+	}
+	return fs.c.Create(path)
+}
+
+func (fs hdfsFS) Open(path string) (io.ReadCloser, error) { return fs.c.Open(fs.Abs(path)) }
+func (fs hdfsFS) ModTime(path string) (time.Time, error)  { return modtime(fs.c.Stat(fs.Abs(path))) }
+func (fs hdfsFS) Size(path string) (int64, error)         { return size(fs.c.Stat(fs.Abs(path))) }
 
 func modtime(fi os.FileInfo, err error) (time.Time, error) {
 	if err != nil {

